homeassistant/components/rainbird/switch.py

```diff
@@ -3,162 +3,98 @@
-            zone_config = discovery_info.get(CONF_ZONES, {}).get(zone, {})
-            time = zone_config.get(CONF_TRIGGER_TIME, discovery_info[CONF_TRIGGER_TIME])
-            name = zone_config.get(CONF_FRIENDLY_NAME)
-            devices.append(
-                RainBirdSwitch(
-                    coordinator,
-                    controller,
-                    zone,
-                    time,
-                    name if name else f"Sprinkler {zone}",
-                )
-            )
-
-    try:
-        await coordinator.async_config_entry_first_refresh()
-    except ConfigEntryNotReady as err:
-        raise PlatformNotReady(f"Failed to load zone state: {str(err)}") from err
-
-    async_add_entities(devices)
-
-    async def start_irrigation(service: ServiceCall) -> None:
-        entity_id = service.data[ATTR_ENTITY_ID]
-        duration = service.data[ATTR_DURATION]
-
-        for device in devices:
-            if device.entity_id == entity_id:
-                await device.async_turn_on(duration=duration)
-
-    hass.services.async_register(
-        DOMAIN,
-        SERVICE_START_IRRIGATION,
-        start_irrigation,
-        schema=SERVICE_SCHEMA_IRRIGATION,
+        for zone in coordinator.data.zones
     )
 
-    async def set_rain_delay(service: ServiceCall) -> None:
-        duration = service.data[ATTR_DURATION]
-
-        await controller.set_rain_delay(duration)
-
-    hass.services.async_register(
-        DOMAIN,
-        SERVICE_SET_RAIN_DELAY,
-        set_rain_delay,
-        schema=SERVICE_SCHEMA_RAIN_DELAY,
+    platform = entity_platform.async_get_current_platform()
+    platform.async_register_entity_service(
+        SERVICE_START_IRRIGATION,
+        SERVICE_SCHEMA_IRRIGATION,
+        "async_turn_on",
     )
 
 
-class RainBirdSwitch(
-    CoordinatorEntity[RainbirdUpdateCoordinator[States]], SwitchEntity
-):
+class RainBirdSwitch(CoordinatorEntity[RainbirdUpdateCoordinator], SwitchEntity):
     """Representation of a Rain Bird switch."""
 
     def __init__(
         self,
-        coordinator: RainbirdUpdateCoordinator[States],
-        rainbird: AsyncRainbirdController,
+        coordinator: RainbirdUpdateCoordinator,
         zone: int,
-        time: int,
-        name: str,
+        duration_minutes: int,
+        imported_name: str | None,
     ) -> None:
         """Initialize a Rain Bird Switch Device."""
         super().__init__(coordinator)
-        self._rainbird = rainbird
         self._zone = zone
-        self._name = name
+        if imported_name:
+            self._attr_name = imported_name
+        else:
+            self._attr_name = f"Sprinkler {zone}"
         self._state = None
-        self._duration = time
-        self._attributes = {ATTR_DURATION: self._duration, "zone": self._zone}
+        self._duration_minutes = duration_minutes
+        self._attr_unique_id = f"{coordinator.serial_number}-{zone}"
+        self._attr_device_info = coordinator.device_info
```

Comment:
Shouldn't each zone be its own device?

https://developers.home-assistant.io/docs/device_registry_index#what-is-a-device

If we do that, I think we need to change the controller service to use the config entry id as id for the controller in the service call.

Reply:
I guess it will still point back to the same config entry, so we could still use a target. Not sure what's more convenient or obvious for users.

Reply:
Ah, I was not aware of that. This makes sense (the valves are separate devices from the controller) and now I see that each zone can be assigned an area which is cool.  The entity names make more sense now, so i've updated that.

Regarding the service: My general impression is that both "device ids" and "config entry ids" are all opaque, though while users may be more comfortable with devices it's all kind of the same. 

The switches now may use the new entity naming scheme. Is changing to entity names considered another breaking change when added at the same time as adding unique ids?(My impression is this will change the entity names.)

I have been exploring the rainbird api more and found a couple extra features like getting the zone names and device names, so we may be able to use those to get even better names. I will send these in follow up PRs.

Reply:
It's best to add unique_id first while keeping the imported names without setting entity name as in my example. For not imported config entries and entities we can set `_attr_has_entity_name` to True. Then after the deprecation period when we remove the import, we can remove the migration code and set `_attr_has_entity_name` generally.